keps/sig-storage/20190408-volume-scheduling-limits.md

```diff
@@ -0,0 +1,294 @@
+    * After deprecation period, scheduler won't schedule any pods that use volumes to a node with missing `CSINode` instance. It is expected that it happens only during node registration when `Node` exists and `CSINode` doesn't and it self-heals quickly.
+
+* `CSINode.status.allocatable` is array of limits. Following combinations are possible:
+
+|  `Volumes` | Description  |
+|  --------- | ------------ |
+|  0         | plugin / CSI driver exists and has zero limit, i.e. can attach no volumes |
+|  X>0       | plugin / CSI driver exists and can attach X volumes (where X > 0) |
+|  X<0       | negative values are blocked by validation
+| Driver is missing in `CSINode.status.allocatable`   | there is no limit of volumes on the node* |
+
+*) This way we are not able to distinguish between a volume plugin / CSI driver not installed on a node or it has been installed and it has no limits.
+  * Predicates modified in this KEP assume that storage provided by **in-tree** volume plugin is available all nodes in the cluster. Other predicate(s) will evaluate `PV.spec.nodeAffinity` and filter out nodes that don't have access to the storage.
+  * For CSI drivers, availability of a CSI driver on a node can be checked in `CSINode.spec`. Its handling is out of scope of this KEP, see non-goals.
+
+CSINode example:
+
+```yaml
+apiVersion: storage.k8s.io/v1beta1
+kind: CSINode
+metadata:
+  name: ip-172-18-4-112.ec2.internal
+spec:
+status:
+  allocatable:
+    # AWS node can attach max. 40 volumes, 1 is reserved for the system
+    - name: ebs.csi.aws.com
+      volumes: 39
+```
+
+* Existing scheduler predicates `MaxEBSVolumeCount`, `MaxGCEPDVolumeCount`, `MaxAzureDiskVolumeCount` and `MaxCinderVolumeCount` are already deprecated.
+  * If any of them is enabled together with `MaxCSIVolumeCountPred`, the deprecated predicate will do nothing (`MaxCSIVolumeCountPred` does the job of counting both in-tree and CSI volumes).
+  * The deprecated predicates will do any useful work only when `MaxCSIVolumeCountPred` predicate is disabled or `CSINode` object does not have limits for a particular driver.
+  * This way, we save CPU by running only one volume limit predicate during deprecation period.
+
+### New API
+
+CSINode gets `Status` struct with `Allocatable`, holding limit of volumes for each volume plugin and CSI driver that can be scheduled to the node.
+
+```go
+type CSINode struct {
+	...
+
+	// spec is the status of CSINode
+	Status CSINodeStatus `json:"status" protobuf:"bytes,3,opt,name=status"`
+}
+
+// CSINodeStatus holds information about the status of all CSI drivers installed on a node
+type CSINodeStatus struct {
+	// allocatable is a list of volume limits for each volume plugin and CSI driver on the node.
+	// +patchMergeKey=name
+	// +patchStrategy=merge
+	Allocatable []VolumeLimits `json:"allocatable" patchStrategy:"merge" patchMergeKey:"name" protobuf:"bytes,1,rep,name=allocatable"`
```

Comment:
can't this be a `map[string]int32` that maps a CSI driver name to its volume limit?

Reply:
We want to add total volume size later. We can add either two maps, `map[string]int32` for count of volumes and `map[string]Quantity` for total size. Or we can have maps of structs as it is now. Both will work, not sure what's the preference here. Also see discussion at https://github.com/kubernetes/enhancements/pull/942#discussion_r276297444